src/Provider/AbstractProvider.php

```diff
@@ -702,18 +702,26 @@ protected function parseResponse(ResponseInterface $response)
      * Custom mapping of expiration, etc should be done here. Always call the
      * parent method when overloading this method.
      *
-     * @param  mixed $result
-     * @return array
+     * @param  array|string $response
+     * @return array|string
+     * @throws UnexpectedValueException
      */
-    protected function prepareAccessTokenResponse(array $result)
+    protected function prepareAccessTokenResponse($response)
```

Comment:
It's possible for `parseResponse()` to return a string without throwing an exception (if it couldn't parse the response as JSON or urlencoded form values). After doing so, the `getResponse()` method (as it is currently named, not as its name changes to in this PR) passes the string to `checkResponse()`. The `checkResponse()` method is implemented by the provider, so the provider could choose to throw an exception if it's not a proper array, or not if they want the string, instead.

As it stands now, `checkResponse()` can return a string value, which gets passed back up the chain and then into the next call to `prepareAccessTokenResponse()`, where it will throw an error for not being an array.

Should `parseResponse()` or `checkResponse()` always ensure the value is an array before passing it back up the chain?